### ISSUE TITLE
[FA3] Precision issue with gated attention (MLA/GLA/MLRA) in Flash Attention Hopper v3.0.0

### ISSUE THREAD

**Issue by SongtaoLiu0823** [Dao-AILab/flash-attention#2372]
**This report is polished with Opus4.6**
# [FA3 Hopper] Precision degradation with gated attention in v3.0.0 (correct in v3.0.0b1)
 
## Summary
 
When using Flash Attention 3 Hopper kernel (v3.0.0) with **gated attention** (as described in [Gated Attention, Qiu et al., NeurIPS 2025](https://arxiv.org/pdf/2505.06708)), we observe significant precision degradation compared to both PyTorch SDPA and FA3 Hopper v3.0.0b1. The issue **only** occurs when gating is enabled; without gating, v3.0.0 matches SDPA precision correctly.
 
## Environment
 
- **Flash Attention version**: 3.0.0 (bug), 3.0.0b1 (correct)
- **GPU**: NVIDIA H100 (Hopper)
- **PyTorch**: 2.x with SDPA as reference
- **Attention variants tested**: MLA (Multi-Head Latent Attention), GLA (Grouped Latent Attention), MLRA (Multi-Head Low-Rank Attention) from [Liu et al., ICLR 2026](https://arxiv.org/pdf/2603.02188)
 
## Problem Description
 
We are training LLMs with latent attention architectures (MLA/GLA/MLRA) augmented with a gating mechanism [Qiu et al., NeurIPS 2025]. When switching from FA3 Hopper v3.0.0b1 to v3.0.0, we observe a noticeable precision gap in training loss, which does **not** appear when gating is disabled.
 
### Key observations
 
| Condition | v3.0.0b1 vs SDPA | v3.0.0 vs SDPA |
|---|---|---|
| **With gating** | ✅ Match | ❌ Mismatch |
| **Without gating** | ✅ Match | ✅ Match |
 
This strongly suggests a regression in v3.0.0 specifically affecting the gating code path in the Hopper kernel.
 
### 1. GLA-2 w/ gating: v3.0.0b1 vs v3.0.0
 
<img width="518" height="258" alt="Image" src="https://github.com/user-attachments/assets/c82a3c4a-c7bf-4d3b-886e-c0b07b2a5bae" />

Training the same GLA-2 model with gating enabled, identical hyperparameters and data:
 
| Step | v3.0.0b1 (validation loss) | v3.0.0 (validation loss) |
|---|---|---|
| 1000 | **3.5550** | **3.6086** |
 
The v3.0.0 loss is notably higher (+0.054), indicating precision issues accumulating during training.
 
### 2. MLRA-2 w/ gating: v3.0.0 vs SDPA (resumed from same checkpoint at step 42000)
 
<img width="2632" height="458" alt="Image" src="https://github.com/user-attachments/assets/8887694a-3a92-4b5f-b6d3-4f08c965c7a2" />
<img width="1983" height="259" alt="Image" src="https://github.com/user-attachments/assets/19fd138a-1273-493b-9127-ed3c699343fc" />

**SDPA reference**:
```
iter 42000: loss 2.2959
iter 42010: loss 2.2818
iter 42020: loss 2.2182
```
 
**FA3 Hopper v3.0.0**:
```
iter 42000: loss 2.3027
iter 42010: loss 2.2878
iter 42020: loss 2.2250
```
 
Per-step loss differences (v3.0.0 − SDPA): `+0.0068`, `+0.0060`, `+0.0068`, showing a consistent positive bias.
 
## Expected Behavior
 
FA3 Hopper v3.0.0 with gating should produce the same precision as v3.0.0b1 and PyTorch SDPA, as it does when gating is disabled.
 
## References
 
- [1] [Multi-Head Low-Rank Attention](https://arxiv.org/pdf/2603.02188), Liu et al., ICLR 2026
- [2] [Gated Attention for Large Language Models: Non-linearity, Sparsity,
and Attention-Sink-Free](https://arxiv.org/pdf/2505.06708), Qiu et al., NeurIPS 2025

**Comment by tridao** [comment]
Can you send a self-contained script that reproduce the error? e.g. construct input tensors, run attn, check output.

**Comment by SongtaoLiu0823** [comment]
```python3
import torch
import torch.nn as nn
import torch.nn.functional as F
import inspect


class GatedAttentionFA3(nn.Module):
    """Gated attention using flash_attn_func (matches actual training model)."""

    def __init__(self, n_embd, n_head, qk_nope_head_dim, qk_rope_head_dim, flash_attn_func):
        super().__init__()
        self.n_head = n_head
        self.qk_nope_head_dim = qk_nope_head_dim
        self.qk_rope_head_dim = qk_rope_head_dim
        self.qk_head_dim = qk_nope_head_dim + qk_rope_head_dim
        self.v_head_dim = qk_nope_head_dim
        self.flash_attn_func = flash_attn_func

        # Projections (same as actual model)
        self.q_proj = nn.Linear(n_embd, n_head * self.qk_head_dim, bias=False)
        self.k_proj = nn.Linear(n_embd, n_head * self.qk_head_dim, bias=False)
        self.v_proj = nn.Linear(n_embd, n_head * self.v_head_dim, bias=False)
        self.gated_proj = nn.Linear(n_embd, n_head * self.v_head_dim, bias=False)
        self.o_proj = nn.Linear(n_head * self.v_head_dim, n_embd, bias=False)

    def forward(self, x):
        """
        x: (B, T, n_embd) — float32 hidden states from embedding
        """
        B, T, _ = x.shape

        # Under autocast, Linear casts float32 input to bf16 internally
        q = self.q_proj(x).view(B, T, self.n_head, self.qk_head_dim)
        k = self.k_proj(x).view(B, T, self.n_head, self.qk_head_dim)
        v = self.v_proj(x).view(B, T, self.n_head, self.v_head_dim)

        # Gating via nn.Linear + sigmoid
        gated_score = F.sigmoid(self.gated_proj(x))  # (B, T, H * D_v)
        gated_score = gated_score.view(B, T, self.n_head, self.v_head_dim)

        # Flash Attention (receives bf16 q/k/v from linear outputs)
        y = self.flash_attn_func(q=q, k=k, v=v, causal=True)

        # Apply gating
        out = gated_score * y
        out = self.o_proj(out.reshape(B, T, -1))
        return out


class GatedAttentionSDPA(nn.Module):
    """Gated attention using F.scaled_dot_product_attention (reference)."""

    def __init__(self, n_embd, n_head, qk_nope_head_dim, qk_rope_head_dim):
        super().__init__()
        self.n_head = n_head
        self.qk_nope_head_dim = qk_nope_head_dim
        self.qk_rope_head_dim = qk_rope_head_dim
        self.qk_head_dim = qk_nope_head_dim + qk_rope_head_dim
        self.v_head_dim = qk_nope_head_dim

        self.q_proj = nn.Linear(n_embd, n_head * self.qk_head_dim, bias=False)
        self.k_proj = nn.Linear(n_embd, n_head * self.qk_head_dim, bias=False)
        self.v_proj = nn.Linear(n_embd, n_head * self.v_head_dim, bias=False)
        self.gated_proj = nn.Linear(n_embd, n_head * self.v_head_dim, bias=False)
        self.o_proj = nn.Linear(n_head * self.v_head_dim, n_embd, bias=False)

    def forward(self, x):
        B, T, _ = x.shape

        q = self.q_proj(x).view(B, T, self.n_head, self.qk_head_dim)
        k = self.k_proj(x).view(B, T, self.n_head, self.qk_head_dim)
        v = self.v_proj(x).view(B, T, self.n_head, self.v_head_dim)

        gated_score = F.sigmoid(self.gated_proj(x))
        gated_score = gated_score.view(B, T, self.n_head, self.v_head_dim)

        # SDPA expects (B, H, T, D)
        y = F.scaled_dot_product_attention(
            q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), is_causal=True
        ).transpose(1, 2)

        out = gated_score * y
        out = self.o_proj(out.reshape(B, T, -1))
        return out


def copy_weights(src, dst):
    """Copy all matching weights from src to dst."""
    with torch.no_grad():
        for name, param in src.named_parameters():
            dst_param = dict(dst.named_parameters())[name]
            dst_param.copy_(param)


def print_diff(label, a, b):
    """Compare two tensors."""
    a_f = a.float()
    b_f = b.float()
    abs_diff = (a_f - b_f).abs()
    rel_diff = abs_diff / (b_f.abs() + 1e-8)

    print(f"\n  [{label}]")
    print(f"    Max  abs diff: {abs_diff.max().item():.6e}")
    print(f"    Mean abs diff: {abs_diff.mean().item():.6e}")
    print(f"    Max  rel diff: {rel_diff.max().item():.6e}")
    print(f"    Mean rel diff: {rel_diff.mean().item():.6e}")


def test_config(flash_attn_func, name, B, T, n_embd, H, qk_nope_head_dim, qk_rope_head_dim):
    D_qk = qk_nope_head_dim + qk_rope_head_dim
    D_v = qk_nope_head_dim

    print(f"\n{'=' * 70}")
    print(f"Config: {name}")
    print(f"  B={B}, T={T}, n_embd={n_embd}, H={H}, D_qk={D_qk}, D_v={D_v}")
    print("=" * 70)

    device = "cuda"
    torch.manual_seed(42)

    # -------------------------------------------------------------------
    # Build two identical modules (shared weights)
    # -------------------------------------------------------------------
    mod_fa3 = GatedAttentionFA3(n_embd, H, qk_nope_head_dim, qk_rope_head_dim, flash_attn_func).to(device)
    mod_sdpa = GatedAttentionSDPA(n_embd, H, qk_nope_head_dim, qk_rope_head_dim).to(device)
    copy_weights(mod_fa3, mod_sdpa)

    # -------------------------------------------------------------------
    # Construct input: float32 hidden states (from nn.Embedding, not
    # downcast by autocast)
    # -------------------------------------------------------------------
    x = torch.randn(B, T, n_embd, device=device, dtype=torch.float32)

    # ===================================================================
    # Test 1: Forward only, WITHOUT gating (pure attention comparison)
    # ===================================================================
    print("\n--- Test 1: Without gating (forward only, mixed precision) ---")
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        # Manually project and run attention without gating
        q = mod_fa3.q_proj(x).view(B, T, H, D_qk)
        k = mod_fa3.k_proj(x).view(B, T, H, D_qk)
        v = mod_fa3.v_proj(x).view(B, T, H, D_v)

        out_fa3 = flash_attn_func(q=q, k=k, v=v, causal=True)
        out_sdpa = F.scaled_dot_product_attention(
            q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), is_causal=True
        ).transpose(1, 2)

    print_diff("No gating (FA3 vs SDPA)", out_fa3, out_sdpa)

    # ===================================================================
    # Test 2: Forward only, WITH gating (full module, mixed precision)
    # ===================================================================
    print("\n--- Test 2: With gating (forward only, mixed precision) ---")
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        out_fa3_g = mod_fa3(x)
        out_sdpa_g = mod_sdpa(x)

    print_diff("With gating (FA3 vs SDPA)", out_fa3_g, out_sdpa_g)

    # ===================================================================
    # Test 3: Forward + backward, WITH gating
    # ===================================================================
    print("\n--- Test 3: Forward + backward with gating (mixed precision) ---")

    x_fa = x.clone().detach().requires_grad_(True)
    x_ref = x.clone().detach().requires_grad_(True)

    # FA3 forward + backward
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        out_fa = mod_fa3(x_fa)
        loss_fa = out_fa.float().sum()
    loss_fa.backward()

    # SDPA forward + backward
    with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
        out_ref = mod_sdpa(x_ref)
        loss_ref = out_ref.float().sum()
    loss_ref.backward()

    print_diff("Forward output", out_fa, out_ref)
    print_diff("dX gradient (hidden states)", x_fa.grad, x_ref.grad)

    for pname in ["q_proj", "k_proj", "v_proj", "gated_proj", "o_proj"]:
        g_fa = getattr(mod_fa3, pname).weight.grad
        g_ref = getattr(mod_sdpa, pname).weight.grad
        if g_fa is not None and g_ref is not None:
            print_diff(f"d{pname}.weight gradient", g_fa, g_ref)

    # ===================================================================
    # Test 4: Multi-step accumulation (simulate gradient_accumulation)
    # ===================================================================
    print("\n--- Test 4: Multi-step gradient accumulation (5 steps) ---")
    n_steps = 5

    mod_fa3.zero_grad()
    mod_sdpa.zero_grad()

    accumulated_loss_fa3 = 0.0
    accumulated_loss_sdpa = 0.0

    for step in range(n_steps):
        torch.manual_seed(100 + step)
        # float32 hidden states (simulating embedding output)
        x_s = torch.randn(B, T, n_embd, device=device, dtype=torch.float32)

        with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
            out_fa = mod_fa3(x_s)
            loss_fa = out_fa.float().sum() / n_steps

            out_ref = mod_sdpa(x_s)
            loss_ref = out_ref.float().sum() / n_steps

        loss_fa.backward()
        loss_ref.backward()

        accumulated_loss_fa3 += loss_fa.item()
        accumulated_loss_sdpa += loss_ref.item()

    print(f"\n  Accumulated loss FA3:  {accumulated_loss_fa3:.6f}")
    print(f"  Accumulated loss SDPA: {accumulated_loss_sdpa:.6f}")
    print(f"  Loss diff:             {accumulated_loss_fa3 - accumulated_loss_sdpa:.6e}")

    print_diff(
        "gated_proj.weight gradient (accumulated)",
        mod_fa3.gated_proj.weight.grad,
        mod_sdpa.gated_proj.weight.grad,
    )
    print_diff(
        "o_proj.weight gradient (accumulated)",
        mod_fa3.o_proj.weight.grad,
        mod_sdpa.o_proj.weight.grad,
    )


def main():
    print("=" * 70)
    print("  Reference: F.scaled_dot_product_attention (SDPA)")
    print("  Mixed precision: torch.autocast(dtype=bfloat16)")
    print("  Input: float32 hidden states (nn.Embedding not downcast)")
    print("  Gating: nn.Linear + sigmoid (same as training code)")
    print("=" * 70)

    # Print version info
    try:
        import flash_attn
        print(f"\nFlash Attention version: {flash_attn.__version__}")
    except Exception:
        print("\nFlash Attention version: unknown")

    print(f"PyTorch version: {torch.__version__}")
    print(f"CUDA version: {torch.version.cuda}")
    print(f"GPU: {torch.cuda.get_device_name(0)}")

    from flash_attn_interface import flash_attn_func

    sig = inspect.signature(flash_attn_func)
    print(f"\nflash_attn_func signature:\n  {sig}\n")

    # ------------------------------------------------------------------
    # Config 1: GLA-2 xxl (matches actual training setup)
    #   n_embd=3072, n_head=24, qk_nope=128, qk_rope=64
    # ------------------------------------------------------------------
    test_config(
        flash_attn_func,
        name="GLA-2 xxl (actual training config)",
        B=2, T=2048, n_embd=3072, H=24,
        qk_nope_head_dim=128, qk_rope_head_dim=64,
    )

    # ------------------------------------------------------------------
    # Config 2: Smaller test for quick iteration
    # ------------------------------------------------------------------
    test_config(
        flash_attn_func,
        name="Small (quick sanity check)",
        B=1, T=512, n_embd=768, H=12,
        qk_nope_head_dim=64, qk_rope_head_dim=32,
    )

    # ------------------------------------------------------------------
    # Config 3: Symmetric head dim (D_qk == D_v)
    # ------------------------------------------------------------------
    test_config(
        flash_attn_func,
        name="Symmetric head dim (D_qk == D_v == 128)",
        B=2, T=2048, n_embd=3072, H=24,
        qk_nope_head_dim=128, qk_rope_head_dim=0,
    )

    print(f"\n{'=' * 70}")
    print("DONE. Compare results between FA3 v3.0.0 and v3.0.0b1.")
    print("Expected: v3.0.0b1 matches SDPA; v3.0.0 shows larger diffs with gating.")
    print("=" * 70)


if __name__ == "__main__":
    main()
```

Here's a self-contained repro script generated by Opus4.6. It builds two identical gated attention modules — one using flash_attn_func, the other using F.scaled_dot_product_attention as reference — with shared weights, and compares their outputs/gradients under torch.autocast(dtype=bfloat16).
The setup matches our training: input hidden states are float32 (from nn.Embedding, not downcast by autocast), then nn.Linear projections produce bf16 q/k/v under autocast, followed by attention and sigmoid gating (gated_score * attn_out).
It runs 4 tests:

1. Forward without gating (expect FA3 ≈ SDPA on both versions)
2. Forward with gating (where we see the v3.0.0 mismatch)
3. Forward + backward with gating (gradient comparison)
4. 5-step gradient accumulation (simulates training drift)

**Comment by tridao** [comment]
No, just the kernel, no training or module plz. Please help identify which is kernel is wrong (fwd or backward, with which inputs).

**Comment by tridao** [comment]
The shorter the script the better

**Comment by SongtaoLiu0823** [comment]
```python3
import torch, torch.nn.functional as F
from flash_attn_interface import flash_attn_func

torch.manual_seed(42)
B, T, H, D, Dv = 2, 2048, 24, 192, 128
q0, k0, v0 = [torch.randn(B,T,H,d, device="cuda", dtype=torch.bfloat16) for d in (D,D,Dv)]
gate = torch.randn(B,T,H,Dv, device="cuda", dtype=torch.bfloat16).sigmoid()
diff = lambda a, b: f"max={(a.float()-b.float()).abs().max().item():.4e}"

# fwd
qf,kf,vf = [x.clone().requires_grad_(True) for x in (q0,k0,v0)]
qr,kr,vr = [x.clone().requires_grad_(True) for x in (q0,k0,v0)]
fa3 = flash_attn_func(q=qf, k=kf, v=vf, causal=True)
ref = F.scaled_dot_product_attention(qr.transpose(1,2), kr.transpose(1,2), vr.transpose(1,2), is_causal=True).transpose(1,2)
print("fwd no_gate:", diff(fa3, ref), " fwd gated:", diff(gate*fa3, gate*ref))

# bwd
(gate * fa3).sum().backward()
(gate * ref).sum().backward()
print("bwd dQ:", diff(qf.grad, qr.grad), " dK:", diff(kf.grad, kr.grad), " dV:", diff(vf.grad, vr.grad))
```

It’s only the **flash_attn_func** kernel. I suspect the issue is in the backward pass: when I remove the gate, there are no loss mismatch, but once I add the gate, the issue appears. So I think the problem is related to the backward computation.

**Comment by tridao** [comment]
What does precision issue mean? What’s the error?

**Comment by SongtaoLiu0823** [comment]
Loss is higher as shown in my experiments.

**Comment by tridao** [comment]
No what's the error / diff in the kernel?

**Comment by SongtaoLiu0823** [comment]
I'm sorry that I didn't check that. I will do some experiments to find the issue. I changed the title to make the issue clearer. Thanks!
